php-frontend/src/test/java/org/sonar/php/cfg/Validator.java

```diff
@@ -0,0 +1,130 @@
+        .withFailMessage(buildDebugMessage("size", debugDotNotation))
+        .hasSize(numberOfBlocks);
+
+    for (ExpectedBlockStructure expected : expectedStructure.values()) {
+      if (expected.isEnd()) {
+        continue;
+      }
+
+      String debugMessage = buildDebugMessage(expected.testId(), debugDotNotation);
+      assertSuccessors(expected, expectedStructure, debugMessage);
+
+      if (testPredecessors) {
+        Set<CfgBlock> expectedPred = getCfgBlocksSet(expected.expectedPredIds(), expectedStructure);
+        Set<CfgBlock> actualPred = expected.actualBlock().predecessors();
+        assertThat(Sets.symmetricDifference(expectedPred, actualPred))
+          .withFailMessage(debugMessage)
+          .isEmpty();
+      }
+    }
+  }
+
+  private static void assertSuccessors(ExpectedBlockStructure expected,
+    Map<String, ExpectedBlockStructure> expectedStructure,
+    String debugMessage) {
+
+    CfgBlock actualBlock = expected.actualBlock();
+
+    if (actualBlock instanceof PhpCfgBranchingBlock) {
+
+      PhpCfgBranchingBlock actualIfBlock = (PhpCfgBranchingBlock) actualBlock;
+      List<CfgBlock> expectedSucc = getCfgBlocksList(expected.expectedSuccIds(), expectedStructure);
+      assertThat(expectedSucc)
+        .withFailMessage(debugMessage)
+        .hasSize(2);
+      assertThat(actualIfBlock.trueSuccessor())
+        .withFailMessage(debugMessage)
+        .isEqualTo(expectedSucc.get(0));
+      assertThat(actualIfBlock.falseSuccessor())
+        .withFailMessage(debugMessage)
+        .isEqualTo(expectedSucc.get(1));
+
+    } else {
+
+      Set<CfgBlock> expectedSucc = getCfgBlocksSet(expected.expectedSuccIds(), expectedStructure);
+      Set<CfgBlock> actualSucc = actualBlock.successors();
+      assertThat(Sets.symmetricDifference(expectedSucc, actualSucc))
+        .withFailMessage(debugMessage)
+        .isEmpty();
+    }
+  }
+
+  private static Set<CfgBlock> getCfgBlocksSet(Collection<String> ids, Map<String, ExpectedBlockStructure> data) {
+    return getCfgBlocksStream(ids, data).collect(Collectors.toSet());
+  }
+
+  private static List<CfgBlock> getCfgBlocksList(Collection<String> ids, Map<String, ExpectedBlockStructure> data) {
+    return getCfgBlocksStream(ids, data).collect(Collectors.toList());
+  }
+
+  private static Stream<CfgBlock> getCfgBlocksStream(Collection<String> ids, Map<String, ExpectedBlockStructure> data) {
+    return ids.stream().map(id -> data.get(id).actualBlock());
+  }
+
+  private static String buildDebugMessage(String blockTestId, String cfgDotNotation) {
+    StringJoiner stringJoiner = new StringJoiner(System.lineSeparator());
+    stringJoiner.add("Not expected CFG structure. There is a problem with block " + blockTestId);
+    stringJoiner.add("Use use a tool like http://www.webgraphviz.com/ to visualize the below graph in dot notation");
+    stringJoiner.add("==========================================");
+    stringJoiner.add(cfgDotNotation);
```

Comment:
I made one test fail:
```
cfg{0[label=" { block( [ id => 3, succ => [4] ] ) } "];1[label=" { block( [ id => 1, succ => [2] ] ) } "];2[label=" { block( [ id => 4, succ => [END] ] ) } "];3[label=" { block( [ id => 2, succ => [3,4] ] ) }  (IF b ) "];4[label="<END>"];5[label=" { block( [ id => 0, succ => [1,END] ] ) }  (IF a ) "];0->2;1->3;2->4;3->0[label=true];3->2[label=false];5->1[label=true];5->3[label=false];}
```

but I can't see graph:
![image](https://user-images.githubusercontent.com/10108186/46670003-4883ca00-cbd1-11e8-954c-239e463e74ce.png)

any idea what's wrong?

Reply:
The content must be inside
```
digraph G {
  "string from error message"
}

```
I'll update the string, it's my bad.